periodic_tasks/meta_vault/tasks.py

```diff
@@ -0,0 +1,142 @@
+import logging
+
+from eth_typing import ChecksumAddress
+from sw_utils import GNO_NETWORKS, convert_to_mgno
+from web3 import Web3
+from web3.types import BlockNumber, HexStr
+
+from periodic_tasks.common.clients import execution_client
+from periodic_tasks.common.contracts import VaultContract, multicall_contract
+from periodic_tasks.common.execution import wait_for_tx_confirmation
+from periodic_tasks.common.graph import graph_get_vaults
+from periodic_tasks.common.networks import ZERO_CHECKSUM_ADDRESS
+from periodic_tasks.common.settings import NETWORK
+from periodic_tasks.common.typings import Vault
+from periodic_tasks.exit.graph import graph_get_exit_requests_by_vaults
+from periodic_tasks.meta_vault.contracts import MetaVaultContract
+from periodic_tasks.meta_vault.graph import graph_get_metavaults
+from periodic_tasks.meta_vault.typings import SubVaultExitRequest
+
+from . import settings
+from .clients import graph_client
+
+logger = logging.getLogger(__name__)
+
+
+async def process_all_meta_vaults(block_number: BlockNumber) -> None:
+    meta_vaults_map = await graph_get_metavaults(settings.META_VAULTS)
+
+    for meta_vault_address, meta_vault in meta_vaults_map.items():
+        logger.info('Processing meta vault: %s', meta_vault_address)
+        await meta_vault_update_state(
+            meta_vault=meta_vault,
+            block_number=block_number,
+        )
+        await process_deposit_to_subvaults(meta_vault_address=meta_vault_address)
+
+
+async def meta_vault_update_state(
+    meta_vault: Vault,
+    block_number: BlockNumber,
+) -> None:
+    # Get sub vaults
+    sub_vaults = await graph_get_vaults(
+        graph_client=graph_client,
+        vaults=meta_vault.sub_vaults,
+    )
+    sub_vaults_to_harvest: list[ChecksumAddress] = []
+    calls: list[tuple[ChecksumAddress, HexStr]] = []
+
+    # Vault contract
+    vault_encoder = VaultContract(
+        abi_path='abi/IEthVault.json',
+        address=ZERO_CHECKSUM_ADDRESS,
+        client=execution_client,
+    ).encoder()
+
+    # Filter harvestable sub vaults and prepare calls for updating their state
+    for sub_vault in sub_vaults.values():
+        if not sub_vault.can_harvest:
+            continue
+
+        sub_vaults_to_harvest.append(sub_vault.address)
+        calls.append(
+            (
+                sub_vault.address,
+                vault_encoder.update_state(
+                    sub_vault.harvest_params,
+                ),
+            )
+        )
+
+    if not sub_vaults_to_harvest:
+        logger.info('No harvestable subvaults for meta vault %s', meta_vault.address)
+        return
+
+    # Collect exit requests for the sub vaults
+    vault_to_exit_requests = await graph_get_exit_requests_by_vaults(
+        vaults=sub_vaults_to_harvest,
```

Comment:
The exit requests must be claimed from all the sub-vaults, not only the ones we plan to harvest.